// ngx-custom-validators/esm5/src/app/less-than/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { lt } from './validator';
/** @type {?} */
var LESS_THAN_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return LessThanValidator; }),
    multi: true
};
var LessThanValidator = /** @class */ (function () {
    function LessThanValidator() {
    }
    /**
     * @return {?}
     */
    LessThanValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = lt(this.lt);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    LessThanValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'lt') {
                this.validator = lt(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    LessThanValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    LessThanValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    LessThanValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[lt][formControlName],[lt][formControl],[lt][ngModel]',
                    providers: [LESS_THAN_VALIDATOR]
                },] }
    ];
    LessThanValidator.propDecorators = {
        lt: [{ type: Input }]
    };
    return LessThanValidator;
}());
export { LessThanValidator };
if (false) {
    /** @type {?} */
    LessThanValidator.prototype.lt;
    /** @type {?} */
    LessThanValidator.prototype.validator;
    /** @type {?} */
    LessThanValidator.prototype.onChange;
}
