// ngx-custom-validators/esm5/src/app/date-iso/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { dateISO } from './validator';
/** @type {?} */
var DATE_ISO_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return DateISOValidator; }),
    multi: true
};
var DateISOValidator = /** @class */ (function () {
    function DateISOValidator() {
    }
    /**
     * @param {?} c
     * @return {?}
     */
    DateISOValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return dateISO(c);
    };
    DateISOValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[dateISO][formControlName],[dateISO][formControl],[dateISO][ngModel]',
                    providers: [DATE_ISO_VALIDATOR]
                },] }
    ];
    return DateISOValidator;
}());
export { DateISOValidator };
