// ngx-custom-validators/esm5/src/app/not-equal/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { notEqual } from './validator';
/** @type {?} */
var NOT_EQUAL_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return NotEqualValidator; }),
    multi: true
};
var NotEqualValidator = /** @class */ (function () {
    function NotEqualValidator() {
    }
    /**
     * @return {?}
     */
    NotEqualValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = notEqual(this.notEqual);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NotEqualValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'notEqual') {
                this.validator = notEqual(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    NotEqualValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    NotEqualValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    NotEqualValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[notEqual][formControlName],[notEqual][formControl],[notEqual][ngModel]',
                    providers: [NOT_EQUAL_VALIDATOR]
                },] }
    ];
    NotEqualValidator.propDecorators = {
        notEqual: [{ type: Input }]
    };
    return NotEqualValidator;
}());
export { NotEqualValidator };
if (false) {
    /** @type {?} */
    NotEqualValidator.prototype.notEqual;
    /** @type {?} */
    NotEqualValidator.prototype.validator;
    /** @type {?} */
    NotEqualValidator.prototype.onChange;
}
