// ngx-custom-validators/esm5/src/app/base64/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { base64 } from './validator';
/** @type {?} */
var BASE64_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return Base64Validator; }),
    multi: true
};
var Base64Validator = /** @class */ (function () {
    function Base64Validator() {
    }
    /**
     * @param {?} c
     * @return {?}
     */
    Base64Validator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return base64(c);
    };
    Base64Validator.decorators = [
        { type: Directive, args: [{
                    selector: '[base64][formControlName],[base64][formControl],[base64][ngModel]',
                    providers: [BASE64_VALIDATOR]
                },] }
    ];
    return Base64Validator;
}());
export { Base64Validator };
