// ngx-custom-validators/esm5/src/app/credit-card/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { creditCard } from './validator';
/** @type {?} */
var CREDIT_CARD_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return CreditCardValidator; }),
    multi: true
};
var CreditCardValidator = /** @class */ (function () {
    function CreditCardValidator() {
    }
    /**
     * @param {?} c
     * @return {?}
     */
    CreditCardValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return creditCard(c);
    };
    CreditCardValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[creditCard][formControlName],[creditCard][formControl],[creditCard][ngModel]',
                    providers: [CREDIT_CARD_VALIDATOR]
                },] }
    ];
    return CreditCardValidator;
}());
export { CreditCardValidator };
