// ngx-custom-validators/esm5/src/app/equal-to/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS, FormControl } from '@angular/forms';
import { equalTo } from './validator';
/** @type {?} */
var EQUAL_TO_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return EqualToValidator; }),
    multi: true
};
var EqualToValidator = /** @class */ (function () {
    function EqualToValidator() {
    }
    /**
     * @return {?}
     */
    EqualToValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = equalTo(this.equalTo);
    };
    /**
     * @param {?} c
     * @return {?}
     */
    EqualToValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    EqualToValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[equalTo][formControlName],[equalTo][formControl],[equalTo][ngModel]',
                    providers: [EQUAL_TO_VALIDATOR]
                },] }
    ];
    EqualToValidator.propDecorators = {
        equalTo: [{ type: Input }]
    };
    return EqualToValidator;
}());
export { EqualToValidator };
if (false) {
    /** @type {?} */
    EqualToValidator.prototype.equalTo;
    /** @type {?} */
    EqualToValidator.prototype.validator;
}
