// ngx-custom-validators/esm5/src/app/range-length/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { rangeLength } from './validator';
/** @type {?} */
var RANGE_LENGTH_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return RangeLengthValidator; }),
    multi: true
};
var RangeLengthValidator = /** @class */ (function () {
    function RangeLengthValidator() {
    }
    /**
     * @return {?}
     */
    RangeLengthValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = rangeLength(this.rangeLength);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    RangeLengthValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'rangeLength') {
                this.validator = rangeLength(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    RangeLengthValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RangeLengthValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    RangeLengthValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[rangeLength][formControlName],[rangeLength][formControl],[rangeLength][ngModel]',
                    providers: [RANGE_LENGTH_VALIDATOR]
                },] }
    ];
    RangeLengthValidator.propDecorators = {
        rangeLength: [{ type: Input }]
    };
    return RangeLengthValidator;
}());
export { RangeLengthValidator };
if (false) {
    /** @type {?} */
    RangeLengthValidator.prototype.rangeLength;
    /** @type {?} */
    RangeLengthValidator.prototype.validator;
    /** @type {?} */
    RangeLengthValidator.prototype.onChange;
}
