// ngx-custom-validators/esm5/src/app/number/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { number } from './validator';
/** @type {?} */
var NUMBER_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return NumberValidator; }),
    multi: true
};
var NumberValidator = /** @class */ (function () {
    function NumberValidator() {
    }
    /**
     * @param {?} c
     * @return {?}
     */
    NumberValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return number(c);
    };
    NumberValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[number][formControlName],[number][formControl],[number][ngModel]',
                    providers: [NUMBER_VALIDATOR]
                },] }
    ];
    return NumberValidator;
}());
export { NumberValidator };
