// ngx-custom-validators/esm5/src/app/array-length/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { arrayLength } from './validator';
/** @type {?} */
var ARRAY_LENGTH_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return ArrayLengthValidator; }),
    multi: true
};
var ArrayLengthValidator = /** @class */ (function () {
    function ArrayLengthValidator() {
    }
    /**
     * @return {?}
     */
    ArrayLengthValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = arrayLength(this.arrayLength);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ArrayLengthValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'arrayLength') {
                this.validator = arrayLength(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    ArrayLengthValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ArrayLengthValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    ArrayLengthValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[arrayLength][formControlName],[arrayLength][formControl],[arrayLength][ngModel]',
                    providers: [ARRAY_LENGTH_VALIDATOR]
                },] }
    ];
    ArrayLengthValidator.propDecorators = {
        arrayLength: [{ type: Input }]
    };
    return ArrayLengthValidator;
}());
export { ArrayLengthValidator };
if (false) {
    /** @type {?} */
    ArrayLengthValidator.prototype.arrayLength;
    /** @type {?} */
    ArrayLengthValidator.prototype.validator;
    /** @type {?} */
    ArrayLengthValidator.prototype.onChange;
}
