// ngx-custom-validators/esm5/src/app/range/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, forwardRef } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { range } from './validator';
/** @type {?} */
var RANGE_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return RangeValidator; }),
    multi: true
};
var RangeValidator = /** @class */ (function () {
    function RangeValidator() {
    }
    /**
     * @return {?}
     */
    RangeValidator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.validator = range(this.range);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    RangeValidator.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var key in changes) {
            if (key === 'range') {
                this.validator = range(changes[key].currentValue);
                if (this.onChange) {
                    this.onChange();
                }
            }
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    RangeValidator.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return this.validator(c);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    RangeValidator.prototype.registerOnValidatorChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    RangeValidator.decorators = [
        { type: Directive, args: [{
                    selector: '[range][formControlName],[range][formControl],[range][ngModel]',
                    providers: [RANGE_VALIDATOR]
                },] }
    ];
    RangeValidator.propDecorators = {
        range: [{ type: Input }]
    };
    return RangeValidator;
}());
export { RangeValidator };
if (false) {
    /** @type {?} */
    RangeValidator.prototype.range;
    /** @type {?} */
    RangeValidator.prototype.validator;
    /** @type {?} */
    RangeValidator.prototype.onChange;
}
